<div style="text-align: center"> <h1> Assignment - 5</div>

<div style="text-align: center"> <h3>CS677 A1 Data Science with Python (Fall 2024)

<div style="text-align: center"> <h4>Submitted by: Aishwarya Malhotra (U17161095)</div>

### **Daytime Trading with Statistical Oracle**

In [1]:
import pandas as pd

In [2]:
df = pd.read_excel("META_weekly_return_detailed.xlsx")

In [3]:
# We sort the dates in accending order
df.sort_values(by = 'Date', inplace = True)

In [4]:
df.drop(columns = ['Label'], inplace = True)

In [5]:
df.head()

Open        High         Low       Close   Adj Close    Volume  \
1257  128.990005  137.509995  128.559998  135.679993  135.272705  28146200   
1256  134.690002  137.169998  131.119995  131.740005  131.344528  22717900   
1255  134.009995  138.000000  133.750000  137.949997  137.535889  29002100   
1254  137.559998  138.869995  135.910004  138.050003  137.635590  20089300   
1253  139.889999  143.139999  139.539993  142.529999  142.102158  26263800   

      Return       Date  Week_Number  Year  Day    Weekday  mean_return  \
1257   0.000 2019-01-02            0  2019    2  Wednesday     0.603333   
1256  -2.904 2019-01-03            0  2019    3   Thursday     0.603333   
1255   4.714 2019-01-04            0  2019    4     Friday     0.603333   
1254   0.072 2019-01-07            1  2019    7     Monday     0.842400   
1253   3.245 2019-01-08            1  2019    8    Tuesday     0.842400   

      volatility  
1257     3.84467  
1256     3.84467  
1255     3.84467  
1254     1.45666  
1253     1.45666

In [6]:
#   Compute the daytime returns as the percent change from 'Open' to 'Close' price
df['daytime_return'] = (df['Close'] - df['Open']) / df['Open'] *100

In [7]:
weekday_returns = df.groupby('Weekday')['daytime_return'].mean().reset_index()


##### **Solution - 1.1**

In [8]:
results_df = (df.groupby(['Year', 'Weekday'])
              .agg(
                  μ_R=('daytime_return', 'mean'),
                  σ_R=('daytime_return', 'std'),
                  count_R_minus=('daytime_return', lambda x: (x < 0).sum()),
                  μ_R_minus=('daytime_return', lambda x: x[x < 0].mean() if not x[x < 0].empty else 0),
                  σ_R_minus=('daytime_return', lambda x: x[x < 0].std() if not x[x < 0].empty else 0),
                  count_R_plus=('daytime_return', lambda x: (x >= 0).sum()),
                  μ_R_plus=('daytime_return', lambda x: x[x >= 0].mean() if not x[x >= 0].empty else 0),
                  σ_R_plus=('daytime_return', lambda x: x[x >= 0].std() if not x[x >= 0].empty else 0)
              )
              .reset_index())

results_df.columns = ['Year', 'Day', 'μ(R)', 'σ(R)', '|R-|', 'μ(R-)', 'σ(R-)', '|R+|', 'μ(R+)', 'σ(R+)']

# Define the custom order for the days of the week
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']

# Convert the 'Day' column to a categorical type with the specified order
results_df['Day'] = pd.Categorical(results_df['Day'], categories=day_order, ordered=True)

# Sort the DataFrame by 'Year' and 'Day'
result_df = results_df.sort_values(by=['Year', 'Day']).reset_index(drop=True)

##### **Solution - 1.2**

In [9]:
for year in result_df['Year'].unique():
    print(f"\nResults for Year {year}:")
    display(result_df[result_df['Year'] == year].drop(columns='Year'))


Results for Year 2019:


Day      μ(R)      σ(R)  |R-|     μ(R-)     σ(R-)  |R+|     μ(R+)  \
0     Monday  0.130572  1.472408    19 -1.151193  1.356257    29  0.970350   
1    Tuesday  0.031175  1.519552    28 -1.049099  1.082071    24  1.291495   
2  Wednesday  0.206974  1.535373    25 -0.949039  0.746789    26  1.318525   
3   Thursday -0.087394  1.157645    24 -1.037584  0.844832    26  0.789705   
4     Friday  0.035728  1.132367    24 -0.881876  0.651324    27  0.851376   

      σ(R+)  
0  0.781903  
1  0.816932  
2  1.248207  
3  0.550648  
4  0.788988


Results for Year 2020:


Day      μ(R)      σ(R)  |R-|     μ(R-)     σ(R-)  |R+|     μ(R+)  \
5     Monday  0.469652  2.025981    18 -1.506050  1.270466    30  1.655073   
6    Tuesday  0.042332  1.837194    25 -1.453792  1.190557    27  1.427633   
7  Wednesday  0.153179  1.975120    25 -1.326489  0.994797    27  1.523242   
8   Thursday  0.067586  1.815286    27 -1.235325  0.878942    25  1.474730   
9     Friday -0.098538  2.063379    26 -1.376135  1.629467    23  1.345702   

      σ(R+)  
5  1.352632  
6  1.080366  
7  1.638347  
8  1.480941  
9  1.470462


Results for Year 2021:


Day      μ(R)      σ(R)  |R-|     μ(R-)     σ(R-)  |R+|     μ(R+)  \
10     Monday  0.317610  1.703564    19 -1.399617  1.002459    28  1.482871   
11    Tuesday -0.063735  1.382080    29 -1.027774  0.945313    23  1.151793   
12  Wednesday -0.252978  1.644795    32 -1.233062  1.171581    20  1.315156   
13   Thursday -0.014118  1.333414    23 -1.085847  1.125823    28  0.866230   
14     Friday  0.166414  1.415591    24 -0.989982  0.811337    26  1.233856   

       σ(R+)  
10  0.889477  
11  0.721792  
12  0.903306  
13  0.691133  
14  0.923722


Results for Year 2022:


Day      μ(R)      σ(R)  |R-|     μ(R-)     σ(R-)  |R+|     μ(R+)  \
15     Monday  0.033266  2.455482    19 -2.194742  1.789698    26  1.661427   
16    Tuesday -0.317643  2.474562    28 -2.015768  1.690838    24  1.663504   
17  Wednesday  0.279551  2.789343    22 -2.210883  1.573810    30  2.105869   
18   Thursday  0.190200  2.849014    26 -1.918497  1.676865    25  2.383244   
19     Friday -0.013492  2.458225    27 -1.857011  1.386965    24  2.060468   

       σ(R+)  
15  1.332649  
16  1.617363  
17  1.932257  
18  2.038171  
19  1.576116


Results for Year 2023:


Day      μ(R)      σ(R)  |R-|     μ(R-)     σ(R-)  |R+|     μ(R+)  \
20     Monday  0.460525  1.364776    17 -1.015549  0.855276    28  1.356712   
21    Tuesday  0.440635  1.520882    22 -0.959830  0.612895    29  1.503056   
22  Wednesday -0.044396  1.676333    25 -1.485120  0.882577    27  1.289608   
23   Thursday  0.198790  2.036113    22 -1.649721  1.390680    29  1.601108   
24     Friday  0.359269  1.842375    24 -1.222823  1.018733    27  1.765572   

       σ(R+)  
20  0.630206  
21  1.067098  
22  0.974275  
23  1.109565  
24  1.115924

##### **Solution - 1.3**

In [10]:
total_negative_days = results_df['|R-|'].sum()
total_positive_days = results_df['|R+|'].sum()

print(f"\nTotal number of days with R−: {total_negative_days}")
print(f"Total number of days with R+: {total_positive_days}")


Total number of days with R−: 600
Total number of days with R+: 658


> **💡Results💡-**
>
> There are more days with non-negative daytime returns compared to negative daytime returns.

##### **Solution - 1.4**

In [11]:
# Overall mean returns for negative and positive returns
mean_negative_return = results_df['μ(R-)'].mean()
mean_positive_return = results_df['μ(R+)'].mean()

print(f"Average loss on down days (μ(R-)): {mean_negative_return}")
print(f"Average gain on up days (μ(R+)): {mean_positive_return}")

if mean_negative_return < mean_positive_return:
    print("The stock loses more on down days than it gains on up days.")
else:
    print("The stock gains more on up days than it loses on down days.")


Average loss on down days (μ(R-)): -1.369312326221264
Average gain on up days (μ(R+)): 1.443532223559222
The stock loses more on down days than it gains on up days.


##### **Solution - 1.5**

**💡Results💡-**

**Volatility across Days:**

No results are not same across days of the week.

While certain days tend to be more volatile and show patterns, there is no consistent behavior across all days of the week, and the difference in returns on down vs up days varies by year and day of the week.

##### **Solution - 2.1**

In [12]:
import pandas as pd

def highlight_max_min(df):
    # Define the columns to highlight
    cols = ['μ(R)', 'σ(R)', 'μ(R-)', 'σ(R-)', 'μ(R+)', 'σ(R+)']

    # Create a style dataframe to store highlighting
    styles = pd.DataFrame('', index=df.index, columns=df.columns)

    for year in df['Year'].unique():
        # Filter the dataframe for the current year
        df_year = df[df['Year'] == year]

        for col in cols:
            # Find index of max and min values
            max_idx = df_year[col].idxmax()
            min_idx = df_year[col].idxmin()

            # Set styles based on the column
            if col in ['σ(R)', 'σ(R-)', 'σ(R+)']:
                # For standard deviation columns, high is lightcoral and low is lightgreen
                styles.loc[max_idx, col] = 'background-color: lightcoral'  # High value
                styles.loc[min_idx, col] = 'background-color: lightgreen'  # Low value
            else:
                # For mean columns, high is lightgreen and low is lightcoral
                styles.loc[max_idx, col] = 'background-color: lightgreen'  # High value
                styles.loc[min_idx, col] = 'background-color: lightcoral'  # Low value

    return styles

# Apply the highlighting function
result_df2 = result_df.style.apply(highlight_max_min, axis=None)
result_df2


**💡Results💡-**

**Return Patterns:**

*   **Mondays** appear to be more favourable often showing highest returns with maximum green highlights (2020, 2021 & 2023). Maybe that's the "Monday Effect" for META. Negative days trends decrease from 2019 to 2023, implying fewer negative days over time.
*   **Tuesday** negative returns are increasing from 2019 to 2022 but in 2023 both negative and positive return averages and show improvement, signaling a potential recovery.
*   **Friday & Thursday** Positive returns recover strongly in 2023. They have the strongest shift from 2020's negative returns to positive returns in 2023.

In [14]:
# Calculate average returns across days of the week
weekly_analysis = result_df.groupby('Day', observed=False).agg({
    'μ(R)': 'mean',
    'μ(R-)': 'mean',
    'μ(R+)': 'mean'
}).reset_index()

# Display the results
weekly_analysis

Day      μ(R)     μ(R-)     μ(R+)
0     Monday  0.282325 -1.453430  1.425286
1    Tuesday  0.026553 -1.301253  1.407496
2  Wednesday  0.068466 -1.440918  1.510480
3   Thursday  0.071013 -1.385395  1.423004
4     Friday  0.089876 -1.265565  1.451395

In [15]:
# Find the best and worst days of the week for each year
best_days = result_df.loc[result_df.groupby('Year')['μ(R)'].idxmax()][['Year', 'Day', 'μ(R)']]
worst_days = result_df.loc[result_df.groupby('Year')['μ(R)'].idxmin()][['Year', 'Day', 'μ(R)']]

# Combine results
best_worst_days = best_days.merge(worst_days, on='Year', suffixes=('_Best', '_Worst'))
best_worst_days

Year   Day_Best  μ(R)_Best  Day_Worst  μ(R)_Worst
0  2019  Wednesday   0.206974   Thursday   -0.087394
1  2020     Monday   0.469652     Friday   -0.098538
2  2021     Monday   0.317610  Wednesday   -0.252978
3  2022  Wednesday   0.279551    Tuesday   -0.317643
4  2023     Monday   0.460525  Wednesday   -0.044396

**💡Interpretation of Table💡-**
On an average Mondays are the best time to invest in META. Wednesday is half and half best for 2019 and 2022 and worst for 2021 and 2023.

##### **Solution - 3.1**

In [16]:
# Calculate average returns across all years
aggregate_analysis = results_df.groupby('Day', observed=False).agg({
    'μ(R)': 'mean'
}).reset_index()

# Determine the best and worst days
best_day = aggregate_analysis.loc[aggregate_analysis['μ(R)'].idxmax()]
worst_day = aggregate_analysis.loc[aggregate_analysis['μ(R)'].idxmin()]

In [17]:
best_day

Day       Monday
μ(R)    0.282325
Name: 0, dtype: object

In [18]:
worst_day

Day      Tuesday
μ(R)    0.026553
Name: 1, dtype: object

##### **Solution - 3.2**

In [19]:
# Calculate overall mean and standard deviation
overall_mean = results_df['μ(R)'].mean()
overall_std = results_df['σ(R)'].mean() 

# Calculate the range
lower_bound = overall_mean - 2 * overall_std
upper_bound = overall_mean + 2 * overall_std

# Count trading days outside the range
outside_range_count = ((df['daytime_return'] < lower_bound) | (df['daytime_return'] > upper_bound)).sum()

# Expected number of days outside the range for 250 trading days
expected_days_outside = 0.05 * 250

# Output results
outside_range_count, expected_days_outside


(74, 12.5)

> **💡Interpretation of Results💡-**
>
> **Observed Count (74 days):** This is significantly higher than the expected count of **12.5 days.**
> **Conclusion:** The observed number of trading days with returns outside the μ ± 2σ range suggests that the  daytime returns are not consistent with the assumption of normality.

##### **Solution - 4.1**

In [20]:
df['Daily Return'] = (df['Close'] - df['Open']) / df['Open']

In [21]:
# Oracle simulation with p = 1
def oracle_simulation(data):
    balance = 100  # Starting with $100
    for index, row in data.iterrows():
        daily_return = (row['Close'] - row['Open']) / row['Open']
        # Oracle predicts correctly every day
        if daily_return > 0:  # Stock price goes up, you invest
            balance *= (1 + row['Daily Return'])
        # If the stock price goes down, the oracle tells you to avoid trading, so no change to balance
    return balance

final_balance_oracle = oracle_simulation(df)
print(f"Final balance after 5 years with Oracle (p=1): ${final_balance_oracle:.2f}")

Final balance after 5 years with Oracle (p=1): $1194731.67


##### **Solution - 4.2**

In [22]:
# Set starting balance and desired amount based on BU ID
bu_id = 95
starting_balance = 100
desired_amount = 100 + bu_id

def days_to_reach_desired_amount(data, starting_balance, desired_amount):
    balance = starting_balance
    day_count = 0
    for index, row in data.iterrows():
        # Update balance
        balance = balance * (1 + row['Daily Return'])
        day_count += 1
        # Check if balance has reached or exceeded desired amount
        if balance >= desired_amount:
            break
    return day_count, balance

# Calculate how many days it takes to reach the desired amount
days, final_balance = days_to_reach_desired_amount(df, starting_balance, desired_amount)

print(f"It took {days} days to reach or exceed the desired amount of ${desired_amount}. Final balance: ${final_balance:.2f}")


It took 1029 days to reach or exceed the desired amount of $195. Final balance: $198.13


##### **Solution - 5.1**

In [23]:
# Function to simulate Buy-and-Hold strategy
def buy_and_hold(data):
    balance = 100  # Starting with $100
    for index, row in data.iterrows():
        # Update balance
        balance *= (1 + row['Daily Return'])
    return balance

final_balance_bh = buy_and_hold(df)
print(f"Final balance after 5 years with Buy-and-Hold: ${final_balance_bh:.2f}")

Final balance after 5 years with Buy-and-Hold: $297.84


##### **Solution - 5.2**

 **💡Comparison to Oracle's Performance💡-**

The oracle's performance dramatically outperforms the buy-and-hold strategy. The oracle's ability to provide accute predictions allows you to grow your initial investment of $100 to over 1 Million dollars, whereas the buy-and-hold strategy leads to a very modest increase.

##### **Solution - 5.3**

In [24]:
import pandas as pd

df['Month'] = pd.to_datetime(df['Date']).dt.month

initial_balance = 100

# Filter out winter months
df_winter_vaca = df[~df['Month'].isin([12, 1, 2])]

# Reset the balance
balance = initial_balance

# Simulate the strategy without winter months
for index, row in df_winter_vaca.iterrows():
    balance *= (1 + row['Daily Return'])

final_balance_winter_vaca = balance
print(f"Final balance after 5 years (excluding winter months): ${final_balance_winter_vaca:.2f}")

Final balance after 5 years (excluding winter months): $173.62


In [25]:
# Summarize findings
results = {
    'Strategy': ['Daytime Buy-and-Hold', 'Daytime Buy-and-Hold with Winter Vacation'],
    'Final Balance ($)': [final_balance_bh, final_balance_winter_vaca]
}

# Convert results to a DataFrame for better visualization
results_df = pd.DataFrame(results)
results_df

Strategy  Final Balance ($)
0                       Daytime Buy-and-Hold         297.836570
1  Daytime Buy-and-Hold with Winter Vacation         173.615726

 **💡Winter - excluding strategy💡-**

The winter - excluding strategy is significantly lower that the full buy-and-hold strategy. In the case, you would have been better off trading year - round.

##### **Solution - 5.4**

In [26]:
import calendar  # Import the calendar module

# Function to simulate Buy-and-Hold with a month excluded
def buy_and_hold_without_month(df, excluded_month):
    balance = 100  # Starting balance
    for index, row in df.iterrows():
        # Skip the rows where the month is the excluded month
        if row['Month'] == excluded_month:
            continue
        # Update balance for all other months
        balance *= (1 + row['Daily Return'])
    return balance

# List of months (1 = January, 12 = December)
months = list(range(1, 13))

# Initialize results dictionary
results = {}

# Full Buy-and-Hold
results['Daytime Buy-and-Hold (Full)'] = buy_and_hold(df) 

# Loop through each month and calculate the balance without that month
for month in months:
    # Use calendar.month_name to get the month name
    strategy_name = f"Daytime B&H without {calendar.month_name[month]}"
    results[strategy_name] = buy_and_hold_without_month(df, month)

results_df = pd.DataFrame(list(results.items()), columns=['Strategy', 'Final Balance'])
results_df

Strategy  Final Balance
0     Daytime Buy-and-Hold (Full)     297.836570
1     Daytime B&H without January     229.384953
2    Daytime B&H without February     265.639528
3       Daytime B&H without March     214.049571
4       Daytime B&H without April     339.940002
5         Daytime B&H without May     212.678942
6        Daytime B&H without June     313.493147
7        Daytime B&H without July     254.006153
8      Daytime B&H without August     268.080129
9   Daytime B&H without September     405.465061
10    Daytime B&H without October     325.522655
11   Daytime B&H without November     243.612992
12   Daytime B&H without December     252.747855

In [27]:
# Identify the best and worst month to take a vacation
best_month = results_df.loc[results_df['Final Balance'].idxmax()]
worst_month = results_df.loc[results_df['Final Balance'].idxmin()]

# Display the results
print(f"\nBest month to take a vacation: {best_month['Strategy']} with a final balance of ${best_month['Final Balance']:.2f}")
print(f"Worst month to take a vacation: {worst_month['Strategy']} with a final balance of ${worst_month['Final Balance']:.2f}")


Best month to take a vacation: Daytime B&H without September with a final balance of $405.47
Worst month to take a vacation: Daytime B&H without May with a final balance of $212.68


##### **Solution - 6.1**

In [28]:
import numpy as np
import pandas as pd

# Function to simulate Oracle with varying p values
def oracle_simulation_with_p(data, p):
    balance = 100  # Start with $100
    for index, row in data.iterrows():
        # Determine if Oracle is correct based on probability p
        oracle_correct = np.random.rand() < p

        # If Oracle predicts correctly AND stock goes up, invest
        if oracle_correct and row['Daily Return'] > 0:
            balance *= (1 + row['Daily Return'])
    return balance

# Define the probability values (0, 0.1, ..., 1)
p_values = np.arange(0, 1.1, 0.1)

# Dictionary to store results for each p value
results = {'p': [], 'Final Balance': []}

# Simulate for each p and store the final balance
for p in p_values:
    final_balance = oracle_simulation_with_p(df, p)
    results['p'].append(p)
    results['Final Balance'].append(final_balance)

# Convert results to a DataFrame for better visualization
results_df = pd.DataFrame(results)

# Format the 'Final Balance' column to display in 2 decimal points
results_df['Final Balance'] = results_df['Final Balance'].map('${:,.2f}'.format)
results_df.loc[len(results_df.index)] = ['Buy-and-Hold', '${:,.2f}'.format(final_balance_bh)]

# Show results DataFrame
results_df

p  Final Balance
0            0.0        $100.00
1            0.1        $199.80
2            0.2        $441.31
3            0.3      $1,624.42
4            0.4      $3,371.71
5            0.5      $9,264.99
6            0.6     $47,707.53
7            0.7     $58,139.59
8            0.8    $192,009.63
9            0.9    $374,146.56
10           1.0  $1,194,731.67
11  Buy-and-Hold        $297.84

 **💡Analysis of Table💡-**

The Buy-and-Hold strategy results in a final balance of $297.84, which is relatively small compared to even a low probability Oracle. This suggests that the **Oracle outperforms**, even at p = 0.1, Buy-and-Hold significantly.

Near p = 0.5: The balance grows significantly, indicating that Oracle performance improves the final balance grows exponentially, indicating strong returns for higher accuracy predictions.

**Moderate Gains with Low p**, the returns are minimal for low values of p (below 0.3).

##### **Solution - 7.1**

In [29]:
import pandas as pd

# Scenario a: Miss the 10 best days
def miss_best_days(df):
    balance = 100
    # Find the 10 best days by sorting by 'Daily Return'
    best_days = df.nlargest(10, 'Daily Return').index
    # Exclude those 10 days and calculate the balance
    df_no_best = df.drop(best_days)
    for index, row in df_no_best.iterrows():
        balance *= (1 + row['Daily Return'])
    return balance

# Scenario b: Invest on the 10 worst days
def invest_worst_days(df):
    balance = 100
    # Find the 10 worst days by sorting by 'Daily Return'
    worst_days = df.nsmallest(10, 'Daily Return').index
    for index, row in df.iterrows():
        if index in worst_days:
            balance *= (1 + row['Daily Return'])  # Invest on worst days
        elif row['Daily Return'] > 0:
            balance *= (1 + row['Daily Return'])  # Invest on other positive days
    return balance

# Scenario c: Miss the 5 best and invest on the 5 worst days
def miss_best_and_invest_worst(df):
    balance = 100
    # Find the 5 best and 5 worst days
    best_days = df.nlargest(5, 'Daily Return').index
    worst_days = df.nsmallest(5, 'Daily Return').index

    for index, row in df.iterrows():
        if index in best_days:
            continue  # Skip the best days
        elif index in worst_days:
            balance *= (1 + row['Daily Return'])  # Invest on worst days (loss)
        elif row['Daily Return'] > 0:
            balance *= (1 + row['Daily Return'])  # Invest on other positive days
    return balance

# Results dictionary
results = {}

# Calculate balances for each strategy
results['Buy-and-hold (B&H)'] = buy_and_hold(df)
results['Daytime B&H (p = 1)'] = oracle_simulation(df)
results['Missed 10 best days'] = miss_best_days(df)
results['Invested during 10 worst days'] = invest_worst_days(df)
results['Missed 5 best & invested during 5 worst days'] = miss_best_and_invest_worst(df)

# Convert results to DataFrame for easy viewing
results_df = pd.DataFrame(list(results.items()), columns=['Strategy', 'Final Balance'])
results_df['Final Balance'] = results_df['Final Balance'].map('${:,.2f}'.format)

# Display the results
results_df

Strategy  Final Balance
0                            Buy-and-hold (B&H)        $297.84
1                           Daytime B&H (p = 1)  $1,194,731.67
2                           Missed 10 best days        $160.24
3                 Invested during 10 worst days    $641,650.33
4  Missed 5 best & invested during 5 worst days    $598,780.11

##### **Solution - 7.2**

 **💡Interpretation of Results: Angry Oracle💡-**

You lose more dramatically by **missing the 10 best daytimes** than by being invested on 10 worst daytimes. Missing out on the most profitable days has a stronger negative impact on your returns.

This suggests that META can experience occational explosive growth wich is critical for the overall portfolio performance.

Interestingly, investing during the worst days still yeilded a considerable balance. This indicates that while those days were suboptimal, they did not lead to complete losses. META has inherent resillience.

##### **Solution - 7.3**

 **💡Angry Oracle vs Statistical Oracle💡-**

The idea is to find the closest match between the final balance from the angry oracle senarios and the statistical oracle's performance at different probabilities p.

(a) **Missed 10 Best Days : Final Balance = `$160.24`**

The closest final balance in the statistical oracle table is between `$100` at p = 0 and `$317.36` at p = 0.1. Based on the closeness to a reasonable estimate for p is approximately p ≈ 0.05.

(b) **Invested During 10 Worst Days : Final Balance = `$641,650.33`**

The closest final balance in the statistical oracle table is between `$383,774.98` at p = 0.9 and `$1,194,731.67` at p = 1. Based on the closeness to a reasonable estimate for p is approximately p ≈ 0.95.


(c) **Missed 5 Best Day & Invested During 5 Worst Days : Final Balance = `$598,780.11`**

The closest final balance in the statistical oracle table is between `$383,774.98` at p = 0.9 and `$1,194,731.67` at p = 1. Based on the closeness to a reasonable estimate for p is approximately p ≈ 0.95.